### ISSUE TITLE
`reportInconsistentConstructor` false positive when `__init__` names a keyword arg covered by `**kwargs` in `__new__`

### ISSUE THREAD

**Issue by thomas-mckay** [DetachHead/basedpyright#1785]
### Description

The following code/[playground](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDKApgDbABQlAxiQIYDO9UAQg0QFzlTdQAmRwKAH0hKIgHcRAChr0ANFABUtEGnrsoYAEYArIlRgLFigNbiVajdr0GAlFAC0APkKlgGgHRfqdRlADCABZIJLxSrPREtpw8fALCQqhIMNKRZEYW6pq6%2BoZQCCq0EBpaYGAkUAC8UAAqIACuREam5qpZ1rn2zlAAcmBint5AA) triggers `reportInconsistentConstructor` on `Child.__init__`, but the signatures are consistent: `param` folds into `**kwargs` in `__new__`, so every valid call to `Child(...)` is accepted by both methods.

```python
from typing import Self

class Base:
    def __new__(cls, *args: object, **kwargs: object) -> Self: ...

class Child(Base):
    def __init__(self, *args: object, param: bool = True, **kwargs: object) -> None: ...
```

**Expected:** no diagnostic — `**kwargs` in `__new__` absorbs all keyword arguments including `param`, so the signatures are structurally consistent at every call site.

**Actual:** `reportInconsistentConstructor` fires on `Child.__init__`.

**Note:** this is the pattern for `__new__`-based mixins that act as transparent pass-throughs, letting subclasses declare specific params in `__init__`. mypy doesn't flag it.

**Comment by DetachHead** [comment]
honestly i can never remember how `__new__` works. is it that the signatures don't need to be identical but rather `__new__`'s parameters just need to be a valid subtype `__init__`'s parameters? @KotlinIsland any idea?

**Comment by KotlinIsland** [comment]
@detachhead init needs to be a subtype of new I think. new is always called, but init isn't always called

**Comment by KotlinIsland** [comment]
Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&reportUnannotatedClassAttribute=false&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDKApgDbABQlAxiQIYDO9UAQg0QFzlTdQAmRwKAH0hKIgHcRAChr0ANFABUtEGnrsoYAEYArIlRgLFigNbiVajdr0GAlFAC0APkKlgGgHRfKPPgOFCqEgw0vRuRhbqmrr6hlAIKrQQGlpgYCRQALxQACogAK5ERqbmqlHWsfbOUAByYGKe3uRCWSxsUgkgSZkARAy8wD225EA)

```python
from typing import Self


class Base:
    def __new__(cls, *args: object, **kwargs: object) -> Self: ...

    def __init__(self, *args: object, param: bool = True, **kwargs: object) -> None: ...  # invalid error

_ = Base(param="asdf")  # correct error
```


i think this is a valid concern, because `__new__` and `__init__` are both invoked from within `type.__call__`, we should consider their call signatures to be combined together into one signature:

```py
class A:
    def __new__(cls, a: int, b: object): ...

    def __init__(cls, a: object, b: int): ...

a.__call__  # (a: int, b: int) -> A
```

**Comment by KotlinIsland** [comment]
i think in general we can say that the signature of the 'constructor' is `__new__ | __init__`

**Comment by DetachHead** [comment]
ok yeah i think we have two options here: 

1. allow mismatching `__init__` and `__new__` parameters, then update basedpyright to verify that the arguments used when instantiating the class are valid for both the `__new__` and `__init__` signatures
2. keep the current logic for instantiation (which just validates the `__init__`) then update `reportInconsistentConstructor` to only report an error if `__new__` isn't a valid subtype of `__init__`

leaning towards option 2 because i think it's likely that mismatching `__new__` and `__init__` is unintentional, so moving the check from the definition to the instantiation can make mistakes go undetected ([like with abstract classes](https://docs.basedpyright.com/latest/benefits-over-pyright/new-diagnostic-rules/#reportimplicitabstractclass))

**Comment by KotlinIsland** [comment]
> mismatching __new__ and __init__ is unintentional

not when there are split classes, like with how `unittest.mock.MagicMock` is typed
